Source/Core/DolphinQt2/Config/ARCodeEditor.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/ARCodeEditor.h"
+
+#include <QDialogButtonBox>
+#include <QFontDatabase>
+#include <QGridLayout>
+#include <QLabel>
+#include <QLineEdit>
+#include <QMessageBox>
+#include <QStringList>
+#include <QTextEdit>
+
+#include "Core/ARDecrypt.h"
+
+ARCodeEditor::ARCodeEditor(ActionReplay::ARCode& ar) : m_code(ar)
+{
+  CreateWidgets();
+  ConnectWidgets();
+
+  m_name_edit->setText(QString::fromStdString(ar.name));
+
+  QString s;
+
+  for (ActionReplay::AREntry& e : ar.ops)
+  {
+    s += QStringLiteral("%1 %2\n")
+             .arg(e.cmd_addr, 8, 16, QLatin1Char('0'))
+             .arg(e.value, 8, 16, QLatin1Char('0'));
+  }
+
+  m_code_edit->setText(s);
+}
+
+void ARCodeEditor::CreateWidgets()
+{
+  m_name_edit = new QLineEdit;
+  m_code_edit = new QTextEdit;
+  m_button_box = new QDialogButtonBox(QDialogButtonBox::Cancel | QDialogButtonBox::Save);
+
+  QGridLayout* grid_layout = new QGridLayout;
+
+  grid_layout->addWidget(new QLabel(tr("Name:")), 0, 0);
+  grid_layout->addWidget(m_name_edit, 0, 1);
+  grid_layout->addWidget(new QLabel(tr("Code:")), 1, 0);
+  grid_layout->addWidget(m_code_edit, 1, 1);
+  grid_layout->addWidget(m_button_box, 2, 1);
+
+  QFont monospace(QFontDatabase::systemFont(QFontDatabase::FixedFont).family());
+
+  m_code_edit->setFont(monospace);
+
+  setLayout(grid_layout);
+}
+
+void ARCodeEditor::ConnectWidgets()
+{
+  connect(m_button_box, &QDialogButtonBox::accepted, this, &ARCodeEditor::accept);
+  connect(m_button_box, &QDialogButtonBox::rejected, this, &QDialog::reject);
+}
+
+void ARCodeEditor::accept()
+{
+  std::vector<ActionReplay::AREntry> entries;
+  std::vector<std::string> encrypted_lines;
+
+  QStringList lines = m_code_edit->toPlainText().split(QStringLiteral("\n"));
+
+  for (int i = 0; i < lines.size(); i++)
+  {
+    QString line = lines[i];
+
+    if (line.isEmpty())
+      continue;
+
+    QStringList values = line.split(QStringLiteral(" "));
+
+    bool good = true;
+
+    u32 addr = 0, value = 0;
+
+    if (values.size() == 2)
+    {
+      addr = values[0].toUInt(&good, 16);
+
+      if (good)
+        value = values[1].toUInt(&good, 16);
+
+      if (good)
+        entries.push_back(ActionReplay::AREntry(addr, value));
+    }
+    else
+    {
+      QStringList blocks = line.split(QStringLiteral("-"));
+
+      if (blocks.size() == 3 && blocks[0].size() == 4 && blocks[1].size() == 4 &&
+          blocks[2].size() == 4)
+      {
+        encrypted_lines.emplace_back(blocks[0].toStdString() + blocks[1].toStdString() +
+                                     blocks[2].toStdString());
+      }
+      else
+      {
+        good = false;
+      }
+    }
+
+    if (!good)
```

Comment:
Perhaps it would be a good idea to split this function up (i.e. the code above this if being in its own function, and below being left here) so the so the sanitizing logic is kept by itself, as opposed to intermixing with other stuff. If this more of a pain than it's worth with how our parsing works currently, feel free to ignore this.

Reply:
I think I'm going to leave it as-is for now.